galaxy/api/base.py

```diff
@@ -0,0 +1,204 @@
+# (c) 2012-2019, Ansible by Red Hat
+#
+# This file is part of Ansible Galaxy
+#
+# Ansible Galaxy is free software: you can redistribute it and/or modify
+# it under the terms of the Apache License as published by
+# the Apache Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# Ansible Galaxy is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# Apache License for more details.
+#
+# You should have received a copy of the Apache License
+# along with Galaxy.  If not, see <http://www.apache.org/licenses/>.
+from django.core.exceptions import PermissionDenied
+from django.http import Http404
+
+from rest_framework import exceptions
+from rest_framework import generics
+from rest_framework import mixins
+from rest_framework.response import Response
+from rest_framework.settings import api_settings
+from rest_framework import status as http_codes
+from rest_framework import views
+
+
+def exception_handler(exc, context):
+    if isinstance(exc, Http404):
+        msg = 'Not found.'
+        data = {'message': str(msg)}
+
+        views.set_rollback()
+        return Response(data, status=http_codes.HTTP_404_NOT_FOUND)
+
+    if isinstance(exc, PermissionDenied):
+        msg = 'Permission denied.'
+        data = {'message': str(msg)}
+
+        views.set_rollback()
+        return Response(data, status=http_codes.HTTP_403_FORBIDDEN)
+
+    if isinstance(exc, exceptions.APIException):
+        headers = {}
+        if getattr(exc, 'auth_header', None):
+            headers['WWW-Authenticate'] = exc.auth_header
+        if getattr(exc, 'wait', None):
+            headers['Retry-After'] = '%d' % exc.wait
+
+        data = {'message': exc.default_detail}
+        errors = []
+
+        if isinstance(exc.detail, list):
+            errors = [{'message': e} for e in exc.detail]
+        elif isinstance(exc.detail, dict):
+            for field, messages in exc.detail.items():
+                if isinstance(messages, str):
+                    messages = [messages]
+                for message in messages:
+                    error = {'message': message}
+                    if field != api_settings.NON_FIELD_ERRORS_KEY:
+                        error['field'] = field
+                    errors.append(error)
+        else:
+            data['message'] = exc.detail
```

Comment:
Worthwhile to also include a `data['detail'] = exc.detail` always?[1]  So the stuff that does have structured details, we always return that as well? Clients like mazer wouldn't use `data['detail']` directly (preferring the `data['messages']` that we have ensured are in a consistent format) but may be useful other tools or troubleshooting.

[1] Assuming exc.detail is serializable

Reply:
See my previous comment. All error details are normalized in DRF exception hierarchy. It can be either `list`, `dict`, `str` or their sub-classes.
All other types are normalized to string representation.
Indeed, there might be recursive structures and we may need to handle this case.

So in this case your proposed `detail` field would be no more than duplicate of `message`.
If we ever need structured details we may extend this API by adding new fields and using custom exception hierarchy handled appropriately in exception handler.

Reply:
That makes sense. I see the `detail` normalizing in the various APIException subclass __init__ now. So while the __init__ might get a more structured `detail` passed in, the instances `.detail` attribute are always normalized (the exception_handler never sees the original `detail` arg passed to the __init__s).